File: 2020/day_1/day_1.go
package main

import "fmt"
import "bufio"
import "os"
import "strconv"

func main() {
    items := make([]int, 0, 100)
    scanner := bufio.NewScanner(os.Stdin)
    for scanner.Scan() {
        newItem, err := strconv.Atoi(scanner.Text())
        if err != nil {
            fmt.Println(err)
        }
        items = append(items, newItem)
    }

    comps, succ := find_complements(items, 3, 2020)

    if !succ {
        fmt.Println("No numbers exist!")
    }

    product := 1
    for i := 0; i < len(comps); i++ {
        product *= comps[i]
    }

    fmt.Println(product)
}

func test_func() []int {
    return []int{1,2,3}
}

func find_complements(list []int, count int, sum int) ([]int, bool) {
    if count == 0 && sum == 0 {
        return []int{}, true
    }
    if count == 0 && sum != 0 {
        return []int{}, false
    }

    for i := 0; i < len(list) - (count - 1); i++ {
        comps, succ := find_complements(list[i+1:], count - 1, sum - list[i])
        if succ {
            return append(comps, list[i]), true
        }
    }

    return []int{}, false
}

File: 2020/day_1/day_1.rs
use std::io::{self, Read};
use std::convert::TryFrom;

fn main() {
    let mut buffer = String::new();

    io::stdin().read_to_string(&mut buffer).expect("Could not read from stdin");

    let mut items: Vec<i64> = Vec::new();
    for line in buffer.split('\n') {
        let mut item = 0;
        match line.parse() {
            Ok(n) => item = n,
            Err(n) => continue,
        }
        items.push(item);
    }

    let comps = find_components(&items, 3, 2020).unwrap();

    let product: i64 = comps.iter().fold(1, |acc, x| acc * x);

    println!("{}", product);
}

fn find_components(items: &[i64], count: i64, sum: i64) -> Option<Vec<i64>> {
    if count == 0 {
        if sum == 0 {
            return Some(Vec::new())
        } else {
            return None
        }
    }

    for tup in items.iter().take(usize::try_from(i64::try_from(items.len()).expect("Long length") - (count - 1)).expect("Too many components")).enumerate() {
        let mut comps;
        match find_components(&items[tup.0 + 1..], count - 1, sum - tup.1) {
            Some(c) => comps = c,
            None => continue,
        }
        comps.push(*tup.1);
        return Some(comps)
    }

    None
}





